server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -220,28 +237,52 @@ public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver refer
         }
 
         if (optionalParams.isEmpty()) {
-            return getDocValueAggregator(reference, DEFAULT_LIMIT);
+            return getDocValueAggregator(reference, DEFAULT_LIMIT, DEFAULT_CAPACITY);
+        }
+
+        if (optionalParams.size() == 1 && optionalParams.getFirst() == null) {
+            return getDocValueAggregator(reference, DEFAULT_LIMIT, DEFAULT_CAPACITY);
         }
 
-        Literal<?> limit = optionalParams.getLast();
-        if (limit == null) {
-            return getDocValueAggregator(reference, DEFAULT_LIMIT);
+        if (optionalParams.size() == 2) {
+            Literal<?> limit = optionalParams.getLast();
+            if (limit != null) {
+                return getDocValueAggregator(reference, (int) limit.value(), DEFAULT_CAPACITY);
+            }
+        }
+
+        if (optionalParams.size() == 3) {
+            int limit, capacity;
+            Literal<?> limitLiteral = optionalParams.get(1);
+            Literal<?> capacityLiteral = optionalParams.get(2);
+
+            if (limitLiteral != null) {
+                limit = (int) limitLiteral.value();
+            } else {
+                limit = DEFAULT_LIMIT;
+            }
```

Comment:
Why is there a default imit in the size() == 3 case, but not in the size == 2 case?

Reply:
👍